pkg/registryserver/registryserver.go

```diff
@@ -115,7 +117,12 @@ func (rs *RegistryServer) RegisterController(ctx context.Context, req *registry.
 
 	if !found {
 		for l := range rs.listeners {
-			l.OnNodeAdded(ctx, node)
+			if err := l.OnNodeAdded(ctx, node); err != nil {
```

Comment:
This is now racy again:
- RegisterController is called twice for the same node
- first call adds the node, fails in OnNodeAdded
- second call finds that the node is registered
- first call removes the node and returns an error
- second call returns success
- node thinks that it is registered but in reality it isn't

The correct solution is to lock the entire RegisterController until it either succeeds (including the OnNodeAdded) or fails.

Injecting foreign code via this callback mechanism is what makes this all so complicated. Can registering nodes and establishing a connection to them be done entirely inside this file?

Reply:
Second `RegisterController `call from same node waits till the first call completes. There is mutex lock based on `NodeId`,

Reply:
Okay, that works.

Can you add explanations of how this locking works to the source code?

Reply:
> Can you add explanations of how this locking works to the source code?

Done